### PR TITLE
error handling implementation

### LINKED ISSUE #22: clean error handling paradigm/semantics
#### What would you like to be added:

We need a cleaner exception/error handling paradigm. Errors that exits the DeFi whale must have a clean serialization and allow deserialization on arrival on the consumer side. Semantics must be clean. Exceptions include validation, 500, bad request, gateway and etc.

#### Why is this needed:

Errors and exception like all things are part of the greater developer quality of life. This is important as it creates a clear and clean semantic for error handling across DeFi services. Error handling like all things, have to scale as our services scale.

An example of a similar spec I wrote a while back: https://github.com/fuxingloh/error-url

### PR DESCRIPTION
**Pull request by fuxingloh**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

A clean exception/error handling paradigm. 

At its root, a response body look like this:
```ts
export interface ApiResponse {
  data?: any
  page?: ApiPage
  error?: ApiError
}
```

any mention of `error` indicates an exception occurred.

```ts
export interface ApiError extends Record<string, any> {
  code: HttpStatus
  type: string
  at: number

  /**
   * Optional message to send out.
   */
  message?: string

  /**
   * Optional injected url that produced ApiError.
   * Injected by exception.interceptor.ts.
   */
  url?: string
}
```

Special errors can have their own property inside ApiException for rich features.

```ts
export class ValidationApiException extends ApiException {
  constructor (errors: ValidationError[]) {
    super({
      code: HttpStatus.UNPROCESSABLE_ENTITY,
      type: ApiErrorType.ValidationError,
      at: Date.now(),
      validation: {
        properties: errors.map(ValidationApiException.mapProperty)
      }
    })
  }

  static mapProperty (error: ValidationError): ApiValidationProperty {
    ...
  }
}
```

error mappings are automatically picked up by `exception.interceptor.ts`.

```ts
export class ExceptionInterceptor implements NestInterceptor {
  map (err: Error): ApiException {
    if (err instanceof ApiException) {
      return err
    }

    if (err instanceof HttpException) {
      return new NestJSApiException(err)
    }

    if (err instanceof JellyfishApiError) {
      return new BadRequestApiException(err.message)
    }

    return new UnknownApiException()
  }
}
```

This allows the controller implementor to throw `ApiException` for rich error, `HttpException` for it to automatically remap into `ApiException`, special errors such as `JellyfishApiError` can be remapped. Lastly unknown errors are ignored, remapped as `UnknownApiException`.

#### Which issue(s) does this PR fixes?:

Fixes #22

### CODE REVIEW

File: __e2e__/call.e2e.ts [hunk 1]
```diff
@@ -40,11 +42,13 @@ it('should 403 with non whitelisted method', async () => {
 
   expect(res.statusCode).toBe(403)
   expect(res.json()).toEqual({
-    error: 'Forbidden',
-    message: 'RPC method not whitelisted',
-    code: 403,
-    url: '/v1/regtest/call/getbalance',
-    at: expect.any(String)
+    error: {
+      code: 403,
```
Comment (jingyi2811) [inline __e2e__/call.e2e.ts:46]: The object sequence of error object should be in the same order as other exception

File: src/module.api/interceptors/api.error.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  url?: string
+}
+
+export class ApiException extends HttpException {
+  protected constructor (spec: ApiError) {
+    super({ error: spec }, spec.code)
+  }
+
+  /**
+   * @param {string} url to inject into ApiException
```
Comment (jingyi2811) [inline src/module.api/interceptors/api.error.ts:45]: Fullstop

File: src/module.api/interceptors/api.error.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    super({
+      code: HttpStatus.INTERNAL_SERVER_ERROR,
+      type: ApiErrorType.UnknownError,
+      at: Date.now()
+    })
+  }
+}
+
+/**
+ * Intercept NestJS exception and map into ApiException
```
Comment (jingyi2811) [inline src/module.api/interceptors/api.error.ts:97]: Fullstop

File: src/module.api/interceptors/api.slice.response.ts [hunk 4]
```diff
@@ -1,11 +1,19 @@
-import { Response } from '@src/module.api/interceptors/response.interceptor'
+import { ApiResponse } from '@src/module.api/interceptors/response.interceptor'
 
 /**
- * SliceResponse, implements Response interface from response.interceptor.
+ * ApiPage for pagination ApiSliceResponse pagination
```
Comment (jingyi2811) [inline src/module.api/interceptors/api.slice.response.ts:4]: Possible full stop or comma

File: src/module.api/interceptors/api.slice.response.ts [hunk 5]
```diff
@@ -74,16 +80,16 @@ export class SliceResponse<T> implements Response {
...
-  static next<T> (data: T[], next?: string): SliceResponse<T> {
-    return new SliceResponse<T>(data, next)
+  static next<T> (data: T[], next?: string): ApiSliceResponse<T> {
+    return new ApiSliceResponse<T>(data, next)
   }
 
   /**
    * @param {T[]} data array slice
    * @param {number} limit number of elements in the data array slice
    * @param {(item: T) => string} nextProvider to get next token when (limit === data array slice)
```
Comment (jingyi2811) [inline src/module.api/interceptors/api.slice.response.ts:90]: Need to add return

File: src/module.api/interceptors/api.error.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    })
+  }
+}
+
+export class NotFoundApiException extends ApiException {
+  constructor (message?: string) {
+    super({
+      code: HttpStatus.NOT_FOUND,
+      type: ApiErrorType.NotFound,
+      at: Date.now(),
```
Comment (jingyi2811) [inline src/module.api/interceptors/api.error.ts:69]: Consistent on object ordering -> code, type, message, at

Comment (fuxingloh) [reply src/module.api/interceptors/api.error.ts:69]: I will follow the interface structure instead.

File: __e2e__/call.e2e.ts (hunk 1 above)
Comment (fuxingloh) [reply __e2e__/call.e2e.ts:46]: don't need to be that strict on tests, it tends to halt progress. For code it's okay.

File: src/module.api/interceptors/api.error.ts (hunk 2 above)
Comment (fuxingloh) [reply src/module.api/interceptors/api.error.ts:45]: The usage of punctuations in the English language aids the reader understanding of the body of text and the information conveyed. Often used to separate written sentences into parts of sentences. And to that effect, they are to separate sentences. In writing; a listicle conveys information that is already separated into parts. `@param` is a form of a listicle.

Not all written work requires a 'period' as they do not carry much information that necessitated the need for the use of punctuation. Documentation in code has similar standards, as they don't carry enough information to form a full sentence. 

`{string} URL to inject into ApiException`
vs
`A string URL is injected into ApiException that will be intercepted by ApiExceptionInterceptor and added into the response body.`

One is a proper sentence and one isn't, this theme of writing is very common in programming documentation as it does not require a full sentence.

**Pull request merged**